packages/tool-cache/src/tool-cache.ts

```diff
@@ -81,24 +97,43 @@ export async function downloadTool(
         throw err
       }
 
+      let error: Error | undefined
       const file: NodeJS.WritableStream = fs.createWriteStream(dest)
       file.on('open', async () => {
-        try {
-          const stream = response.message.pipe(file)
-          stream.on('close', () => {
-            core.debug('download complete')
-            resolve(dest)
-          })
-        } catch (err) {
-          core.debug(
-            `Failed to download from "${url}". Code(${response.message.statusCode}) Message(${response.message.statusMessage})`
-          )
-          reject(err)
-        }
+        const responseMessageFactory = getGlobal<() => stream.Readable>(
+          'TEST_DOWNLOAD_TOOL_RESPONSE_MESSAGE_FACTORY',
+          () => response.message
+        )
+        const readStream = responseMessageFactory()
+        readStream.pipe(file)
+        readStream.on('error', (err: Error) => {
+          file.end()
```

Comment:
per docs, when piping error on the read stream does not close the write stream. needs to be called manually